### DISCUSSION TITLE
New user; Why is zppy looking for .sgs.h0 file

### DISCUSSION THREAD

**Discussion by evasinha-pnnl** [E3SM-Project/zppy#305] [Questions] [answered]
I am a new user to zppy and am using the following configuration file on compy to create lat_lon and timeseries plot of land only simulations.

```
[default]
input = /compyfs/sinh210/e3sm_scratch/20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST
input_subdir = run
output = /compyfs/sinh210/e3sm_scratch/20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST
case = 20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST
www = /compyfs/www/sinh210/E3SM/v2/global/trans
partition = short
model_name = elm

[climo]
active = True
input_files = 'elm.h0'
mapping_file = /qfs/people/zender/data/maps/map_r05_to_cmip6_180x360_aave.20200901.nc
vars = "TLAI,EFLX_LH_TOT,GPP,NPP,NBP,QSOIL,QVEGT,QVEGE,QRUNOFF,TOTVEGC,TOTSOMC,TOTSOMN,TOTSOMP,FPG,FPG_P,TBOT,FSDS,RAIN,CPOOL,SR,RH2M"
years = "1995:2014:20",

[glb]
active = True
weights = 'area'
input_files = 'elm.h0'
mapping_file = /qfs/people/zender/data/maps/map_r05_to_cmip6_180x360_aave.20200901.nc
vars = "TLAI,EFLX_LH_TOT,GPP,NPP,NBP,QSOIL,QVEGT,QVEGE,QRUNOFF,TOTVEGC,TOTSOMC,TOTSOMN,TOTSOMP,FPG,FPG_P,TBOT,FSDS,RAIN,CPOOL,SR,RH2M"
years = "1995:2014:20",

[ts]
active = True
mapping_file = /qfs/people/zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc
vars = "TLAI,EFLX_LH_TOT,GPP,NPP,NBP,QSOIL,QVEGT,QVEGE,QRUNOFF,TOTVEGC,TOTSOMC,TOTSOMN,TOTSOMP,FPG,FPG_P,TBOT,FSDS,RAIN,CPOOL,SR,RH2M"
years = "1850:2014:12",

[e3sm_diags]
active = False

[e3sm_diags_vs_model]
active = False

[amwg]
active = False

[mpas_analysis]
active = False
```

I am getting the following error messages:

1. `ncclimo: ERROR Unable to find required input file /compyfs/sinh210/e3sm_scratch/20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST/run/20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.sgs.h0.1995-01.nc`. Why is zppy looking for sgs file instead of elm?

2. `ts_only
ncclimo: ERROR The 12 files specified (via stdin pipe, positional, or input directory) are expected to contain 1 years of data whereas the date options specify 12 years of data`. How can I modify my cfg to address this mistake?

**Comment by forsyth2** [comment]
I am able to reproduce the errors with a script modified to my directory. (Note that error 1 appears in the `climo*.o*` file and that error 2 appears in the `ts.*o.*` files). I will try to debug.

```
[default]
input = /compyfs/sinh210/e3sm_scratch/20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST
input_subdir = run
output = /compyfs/fors729/n305.20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST
case = 20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST
www = /compyfs/www/fors729/zppy_support
partition = short
model_name = elm

[climo]
active = True
input_files = 'elm.h0'
mapping_file = /qfs/people/zender/data/maps/map_r05_to_cmip6_180x360_aave.20200901.nc
vars = "TLAI,EFLX_LH_TOT,GPP,NPP,NBP,QSOIL,QVEGT,QVEGE,QRUNOFF,TOTVEGC,TOTSOMC,TOTSOMN,TOTSOMP,FPG,FPG_P,TBOT,FSDS,RAIN,CPOOL,SR,RH2M"
years = "1995:2014:20",

[glb]
active = True
weights = 'area'
input_files = 'elm.h0'
mapping_file = /qfs/people/zender/data/maps/map_r05_to_cmip6_180x360_aave.20200901.nc
vars = "TLAI,EFLX_LH_TOT,GPP,NPP,NBP,QSOIL,QVEGT,QVEGE,QRUNOFF,TOTVEGC,TOTSOMC,TOTSOMN,TOTSOMP,FPG,FPG_P,TBOT,FSDS,RAIN,CPOOL,SR,RH2M"
years = "1995:2014:20",

[ts]
active = True
mapping_file = /qfs/people/zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc
vars = "TLAI,EFLX_LH_TOT,GPP,NPP,NBP,QSOIL,QVEGT,QVEGE,QRUNOFF,TOTVEGC,TOTSOMC,TOTSOMN,TOTSOMP,FPG,FPG_P,TBOT,FSDS,RAIN,CPOOL,SR,RH2M"
years = "1850:2014:12",

[e3sm_diags]
active = False

[e3sm_diags_vs_model]
active = False

[amwg]
active = False

[mpas_analysis]
active = False
```

**Reply by forsyth2** [reply]
Re: `sgs` -- this is the default for non-atmosphere. See:
- https://github.com/E3SM-Project/zppy/blob/main/zppy/templates/ts.bash#L94 -- tested with https://github.com/E3SM-Project/zppy/blob/main/tests/integration/template_complete_run.cfg#L51
- https://github.com/E3SM-Project/zppy/blob/main/zppy/templates/climo.bash#L51 -- we don't seem to have tested `climo` with land before.

I ran with `debug = True` and  got the following, which wasn't particularly enlightening for me.

`climo`:
```
+ ncclimo --case=20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST --jobs=4 --thr=1 --vars=TLAI,EFLX_LH_TOT,GPP,NPP,NBP,QSOIL,QVEGT,QVEGE,QRUNOFF,TOTVEGC,TOTSOMC,TOTSOMN,TOTSOMP,FPG,FPG_P,TBOT,FSDS,RAIN,CPOOL,SR,RH2M --parallel=mpi --yr_srt=1995 --yr_end=2014 --input=/compyfs/sinh210/e3sm_scratch/20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST/run --map=/qfs/people/zender/data/maps/map_r05_to_cmip6_180x360_aave.20200901.nc --output=trash --regrid=output --prc_typ=sgs
[...]
ncclimo: ERROR Unable to find required input file /compyfs/sinh210/e3sm_scratch/20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST/run/20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.sgs.h0.1995-01.nc
ncclimo: HINT All files implied to exist by the climatology bounds (start/end year/month) and by the specified (with -P or -m) or default model type, must be in /compyfs/sinh210/e3sm_scratch/20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST/run before ncclimo will proceed
+ '[' 1 '!=' 0 ']'
+ cd /compyfs/fors729/n305.20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST/post/scripts
+ echo 'ERROR (2)'
```

`ts`:
```
+ ls '20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.eam.h0.1850-*.nc' '20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.eam.h0.1851-*.nc' '20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.eam.h0.1852-*.nc' '20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.eam.h0.1853-*.nc' '20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.eam.h0.1854-*.nc' '20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.eam.h0.1855-*.nc' '20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.eam.h0.1856-*.nc' '20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.eam.h0.1857-*.nc' '20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.eam.h0.1858-*.nc' '20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.eam.h0.1859-*.nc' '20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.eam.h0.1860-*.nc' '20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.eam.h0.1861-*.nc'
+ ncclimo -c 20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST -v TLAI,EFLX_LH_TOT,GPP,NPP,NBP,QSOIL,QVEGT,QVEGE,QRUNOFF,TOTVEGC,TOTSOMC,TOTSOMN,TOTSOMP,FPG,FPG_P,TBOT,FSDS,RAIN,CPOOL,SR,RH2M --mem_mb=0 --split --yr_srt=1850 --yr_end=1861 --ypf=12 --map=/qfs/people/zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc -o trash -O output --prc_typ=eam
+ cat input.txt
ncclimo: ERROR The 12 files specified (via stdin pipe, positional, or input directory) are expected to contain 1 years of data whereas the date options specify 12 years of data
ncclimo: HINT Number of files at 12 files-per-year must match number of years implied by arguments to start- and end-year switches (--yr_srt=1850 and --yr_end=1861, respectively). Check that the stdin pipe, positional arguments, or globbed input directory contain the intended number of filenames.
+ '[' 1 '!=' 0 ']'
+ cd /compyfs/fors729/n305.20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST/post/scripts
+ echo 'ERROR (1)'
+ exit 1
```

I tried changing the `ts` year increment from 12 to 10, and got `HINT Provide input filenames in multiples of 12`, which has historically showed up when something else was wrong (see #125) -- notably when the case name and input path were inconsistent. However, your case name and input path do match up and you're already getting a specific error about a missing input file. 

Most of my experience is with running water cycle simulations, so it may be that I'm unfamiliar with how land data is handled/organized. For example, `/compyfs/fors729/E3SMv2/v2.LR.historical_0201/archive/lnd/hist` has a `v2.LR.historical_0201.elm.h0.YYYY-MM.nc` file for each month and each year. However, your `/compyfs/sinh210/e3sm_scratch/20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST/run` directory seems to have significantly fewer files -- only a `0220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.elm.h0.YYYY-02-01-00000.nc` for each year.

@chengzhuzhang This is on Compy -- could either of the errors be related to the Compy remapping issue?
@golaz Let me know if I'm missing something here or if you have any insights into what is happening.

Thanks!

**Reply by evasinha-pnnl** [reply]
For each year, I have one *.h0* that has monthly outputs for 12 months.

**Reply by forsyth2** [reply]
For `ts`, `zppy` concatenates the input files into one file (https://github.com/E3SM-Project/zppy/blob/main/zppy/templates/ts.bash#L62) and sends that one file into `ncclimo`. So, running from your directory, I see 12 input files (1850-02-01 ... 1861-02-01) listed as being concatenated together.

I'm wondering if `ncclimo: ERROR The 12 files specified (via stdin pipe, positional, or input directory) are expected to contain 1 years of data whereas the date options specify 12 years of data` means "for 12 years of data (1850-1861), we would have expected 12 years * 12 months/year * 1 month/file = 144 files, not 12." But as far as `ncclimo` is concerned, it was only passed one input file (the concatenation), so I'm not sure what's going on. I tried setting `--days_per_file = 365` (http://nco.sourceforge.net/nco.html#ncclimo) but that didn't seem to make a difference.

@czender Do you have any suggestions as to why the `ERROR The 12 files specified (via stdin pipe, positional, or input directory) are expected to contain 1 years of data whereas the date options specify 12 years of data` is occurring? Thanks!

**Reply by forsyth2** [reply]
Re: `ncclimo: ERROR Unable to find required input file /compyfs/sinh210/e3sm_scratch/20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST/run/20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.sgs.h0.1995-01.nc` -- even if I manually change `sgs` to `elm`, the file name pattern doesn't match up with the `20220726_global_f09_g16_I1850GSWELMCNCROP_trans_HIST.elm.h0.1995-02-01-00000.nc` that's actually in your directory. I'm looking at the `climo` bash file and I don't see anywhere where the `1995-01` file name gets passed in, so I think it may be inferred by `ncclimo`.

**Reply by czender** [reply]
Hi All,
Thank you for making me aware of this issue. I am 9 timezones away and can more fully help next week. I have some thoughts about what is happening and why. First, the symptom that the input files are missing is clearly due to the creation of slightly incorrect input filenames probably by `ncclimo` though possibly by `zppy` (not sure yet), need to test. Once I find out this should be a straightforward fix.  While the intended way to invoke `ncclimo` for ELM is `ncclimo -P elm`, for a while we had to invoke it as `ncclimo -P sgs` to workaround an earlier `ncclimo` bug that was, IIRC, in E3SMU 1.5. I suspect that's where the extra `sgs` in the constructed filename comes from. I finally received renewed access to my suspended Compy account last week and will investigate this soon.

**Reply by czender** [reply]
I'm currently looking into this, and may be interrupted by traveling through 3 airports before I finish...
The "good news" is that I too can reproduce the problem(s). The workaround for one issue with the climos is to add the `--mdl_nm=clm` option to the run command that @forsyth2 shows above. This causes `ncclimo` to look for `elm.h0` not `sgs.h0` files. A different workaround is to use `-P elm`  instead of `-P sgs` but that will only work with E3SMU 1.7 not 1.6 (which for now  is the default on Compy). Unfortunately, neither of those workarounds is sufficient because, as `ncclimo` prints to its standard output, `Each input file assumed to contain statistics for one month`. @evasinha your files contain 12 months not 1 month. Allowing `ncclimo` to handle that is on the list but will take time. @forsyth2 good job deducing this and yes, that is why those error messages occur for the `climo `command.

The `ts` command above fails for similar reasons. First there is a typo (change `eam` to `elm`) in that command. Then it will break because there are 12 months, not 1 month, per input file. @forsyth2 it is imprecise to say the `zppy` concatenates the files together and sends those to `ncclimo`. It is more accurate to say the zppy concatenates the _**filenames**_ together into a single files, and sends that to `ncclimo`. 

One way to proceed is to split the input files from yearly into monthly files, then re-run the commands with the supplied workarounds. And @evasinha you may want to output one file per month from ELM instead of one per year until `ncclimo` supports that feature.

**Reply by evasinha-pnnl** [reply]
@czender Thanks for examining the error in more detail. Yes, know that I understand what is causing the issue with zppy better, I will output one file per month for all future runs.

**Reply by forsyth2** [reply]
Thank you @czender!

**Reply by czender** [reply]
FYI I have fixed (in the latest snapshot) the problem where specifying `-P sgs` results in an unwanted `sgs` in the expected input filenames. However, the inability to handle 12 months per file is still there.

**Reply by evasinha-pnnl** [reply]
Thank you @czender. I should be able to bypass the 12 months per file issue with the new runs I am generating that outputs one file per month. Has your recent fix regarding `-P sgs` merged into the main branch? How do I access the fix?

**Comment by czender** [comment, marked as answer]
It's in the main branch of NCO. There are three ways to access it: 1. Clone and build NCO from source and call `ncclimo` yourself. That is doable though not trivial. 2. Copy just the ncclimo executable from ~zender/bin/ncclimo to someplace that has precedence over the E3SMU ncclimo. 3. Use the NCO in my personal directories on Compy. You would have to first do something like
```
source /share/apps/intel/2020/compilers_and_libraries_2020.0.166/linux/bin/compilervars.sh intel64 # Compy
module purge
module load intel/20.0.0
module load intelmpi/2020 # Required for ncclimo/ncremap in MPI-mode
module load netcdf/4.6.3
export PATH="~zender/bin:${PATH}"
export LD_LIBRARY_PATH="~zender/lib:${LD_LIBRARY_PATH}"
```

**Reply by evasinha-pnnl** [reply]
Thanks @czender

**Reply by evasinha-pnnl** [reply]
@czender: I logged into compy and the had the following modules loaded by default 
```module list
Currently Loaded Modulefiles:
 1) ncl/6.6.2  
```

I then pasted: 
```source /share/apps/intel/2020/compilers_and_libraries_2020.0.166/linux/bin/compilervars.sh intel64 # Compy
module purge
module load intel/20.0.0
module load intelmpi/2020 # Required for ncclimo/ncremap in MPI-mode
module load netcdf/4.6.3
export PATH="~zender/bin:${PATH}"
export LD_LIBRARY_PATH="~zender/lib:${LD_LIBRARY_PATH}"

module list
Currently Loaded Modulefiles:
 1) intel/20.0.0   2) intelmpi/2020   3) netcdf/4.6.3(default)
```

followed by:
```ncclimo \
> --case=20220909_global_f09_g16_I20TRGSWELMCNCROP_trans_HIST \
> --jobs=4 \
> --thr=1 \
> --vars=TLAI,EFLX_LH_TOT,GPP,NPP,NBP,QSOIL,QVEGT,QVEGE,QRUNOFF,TOTVEGC,TOTSOMC,TOTSOMN,TOTSOMP,FPG,FPG_P,TBOT,FSDS,RAIN,CPOOL,SR,RH2M \
> --parallel=mpi \
> --yr_srt=1905 \
> --yr_end=1924 \
> --input=/compyfs/sinh210/e3sm_scratch/20220909_global_f09_g16_I20TRGSWELMCNCROP_trans_HIST/run \
> --map=/qfs/people/zender/data/maps/map_r05_to_cmip6_180x360_aave.20200901.nc \
> --output=trash \
> --regrid=output \
> --prc_typ=sgs
```

Please note that the output folder is different. This is new run directory that outputs one file per month. I still got the following error message:
```ncclimo: WARNING ncclimo CLM/ELM output typically uses sub-gridscale (SGS) fractions that can only be properly renormalized if the SGS variable option to ncclimo is invoked, but that option was not invoked
ncclimo: HINT Provide ncclimo with the SGS fraction variable using the "--sgs_frc=${sgs_frc_nm}" option, otherwise the regridded data will likely not be properly or conservatively regridded
ncclimo: ERROR MPI master process unable to find node-list for distributing jobs
ncclimo: ncclimo uses first node-list found in $COBALT_NODEFILE, $PBS_NODEFILE, or $SLURM_NODELIST
ncclimo: However, none of these environment variables are set so there is no node-list for distributing MPI jobs
ncclimo: HINT: Requesting MPI-parallelism (i.e., invoking ncclimo with "-p mpi") in a non-MPI environment will trigger this error. Use "-p mpi" only when one of the preceding schedulers has allocated (for interactive use) or will allocate (for non-interactive use) the compute nodes. Otherwise use the default background parallelism (use "-p bck" or omit the option completely) or use serial mode (use "-p serial"). See http://nco.sf.net/nco.html#par_typ for more information on parallelism.
```

**Reply by czender** [reply]
That is actually encouraging. please try these two changes (simultaneously) 
1. `--parallel=bck` (instead of MPI)
2. `--prc_typ=elm` (instead of sgs)

**Reply by evasinha-pnnl** [reply]
Thank you @czender. I modified the two lines as per your suggestion and removed the `--map` and `--regrid` option:
```
ncclimo \
--case=20220909_global_f09_g16_I20TRGSWELMCNCROP_trans_HIST \
--jobs=4 \
--thr=1 \
--vars=TLAI,EFLX_LH_TOT,GPP,NPP,NBP,QSOIL,QVEGT,QVEGE,QRUNOFF,TOTVEGC,TOTSOMC,TOTSOMN,TOTSOMP,FPG,FPG_P,TBOT,FSDS,RAIN,CPOOL,SR,RH2M \
--parallel=bck \
--yr_srt=1905 \
--yr_end=1924 \
--input=/compyfs/sinh210/e3sm_scratch/20220909_global_f09_g16_I20TRGSWELMCNCROP_trans_HIST/run \
--output=/compyfs/sinh210/e3sm_scratch//20220909_global_f09_g16_I20TRGSWELMCNCROP_trans_HIST/climo/ \
--prc_typ=elm
```

and was able to successfully generate the climatology files :).

**Reply by czender** [reply]
Yay! You shouldn't have needed to remove the `map` and `regrid` options, though that's up to you. You can close this issue unless there is anything that still does not work for you (and does not require a new feature to fix).

**Reply by forsyth2** [reply]
> You can close this issue

This is a discussion page, not an issue, so it can't be "closed." There is an option to delete the discussion, but I would prefer discussion pages to stay up to serve as a FAQ for `zppy`.

**Reply by czender** [reply]
Thanks, I learned something. Didn't realize GitHub even had Discussion pages. Yes, by all means keep it don't delete it.

**Reply by forsyth2** [reply]
@czender @evasinha @golaz I want to make sure I'm getting the right take-aways from this thread:

Issue 1: `climo` -- Why is zppy looking for sgs file instead of elm?
-

`sgs` is default -- e.g., https://github.com/E3SM-Project/zppy/blob/main/zppy/templates/climo.bash#L51. Note that this works fine for land `ts` -- https://github.com/E3SM-Project/zppy/blob/main/zppy/templates/ts.bash#L94; we have a `land_monthly` `ts` subtask in our testing: https://github.com/E3SM-Project/zppy/blob/main/tests/integration/template_complete_run.cfg#L51.

> While the intended way to invoke `ncclimo` for ELM is `ncclimo -P elm`, for a while we had to invoke it as `ncclimo -P sgs` to workaround an earlier `ncclimo` bug that was, IIRC, in E3SMU 1.5.

>  The workaround for one issue with the climos is to add the `--mdl_nm=clm` option to the run command 

> A different workaround is to use `-P elm` instead of `-P sgs` but that will only work with E3SMU 1.7 not 1.6 (which for now is the default on Compy)

Does this mean the `sgs` references in `climo.bash` and `ts.bash` should be replaced by `elm`? Or at the least, there should be a check `input_files.split(".")[0] == 'elm'` in addition to the existing  `input_files.split(".")[0] == 'eam'`? And in the `elm` case set `--mdl_nm=clm` or `-P elm` instead of `--prc_typ=sgs`? **If so, I can make a GitHub issue for that.**

> your files contain 12 months not 1 month. Allowing `ncclimo` to handle that is on the list but will take time.

If `ncclimo` doesn't have that functionality, is there anything `zppy` can do as a workaround?

Issue 2: `ts` -- `expected to contain 1 years of data whereas the date options specify 12 years of data`
-

> First there is a typo (change `eam` to `elm`) in that command

I don't think that is a typo. I think that is because nothing in the `[ts]` task specifies it is for land, not atmosphere. https://github.com/E3SM-Project/zppy/blob/main/tests/integration/template_complete_run.cfg#L54 specifies `input_files = "elm.h0"` for instance.

> Then it will break because there are 12 months, not 1 month, per input file.

> It is more accurate to say the `zppy` concatenates the _**filenames**_ together into a single files, and sends that to `ncclimo`.

So, `ncclimo` is actually getting a list of 12 (one-per-month) filenames in the usual case? And it failed in this case because it was getting 1 (12 months) filename? 

> One way to proceed is to split the input files from yearly into monthly files, then re-run the commands with the supplied workarounds.

Is there a command `zppy` could include to do that automatically for the user?  **If so, I can make a GitHub issue for that.**

**Reply by evasinha-pnnl** [reply]
@forsyth2 Yes, all identified GitHub issues look correct to me.

**Reply by czender** [reply]
> Does this mean the sgs references in climo.bash and ts.bash should be replaced by elm? 

When `zppy` uses E3SMU 1.7+ then it should invoke `ncclimo` with `-P model_name`. That would simplify both `climo.bash` and `ts.bash`. Note that `model_name` is then used to construct the filenames. So if the model_name is `elm` and the filenames are v1-style and thus use `clm2`, then `ncclimo` needs an additional argument (`--mdl_nm=clm2`) to construct the filenames correctly.

When invoked with `-P model_name`, `ncclimo` itself will internally generate the correct SGS settings for the models (ELM, MPAS-Seaice) that require SGS. This is probably more robust than `zppy` trying to generate the SGS settings.

> Or at the least, there should be a check input_files.split(".")[0] == 'elm' in addition to the existing input_files.split(".")[0] == 'eam'? And in the elm case set --mdl_nm=clm or -P elm instead of --prc_typ=sgs?

Something like this (with no `else` clause) should work in all cases:

```
{%- if input_files.split(".")[0] == 'cam' or  input_files.split(".")[0] == 'eam' or input_files.split(".")[0] == 'elm' or input_files.split(".")[0] == 'clm2'  %}
--prc_typ={{ input_files.split(".")[0] }}
```
Of course I recommend testing that to be sure answers are BFB as expected.

>  If so, I can make a GitHub issue for that.

Sounds good.

> If ncclimo doesn't have that functionality, is there anything zppy can do as a workaround?

If the input files contain multiple months rather than single months then `zppy` could split them up into monthly files as described above. However, if the first month is not January and the last month is not December then `ncclimo` would need to be invoked with the correct `--mth_srt` and `--mth_end` options as well.

> So, ncclimo is actually getting a list of 12 (one-per-month) filenames in the usual case? And it failed in this case because it was getting 1 (12 months) filename?

Yes and yes

> Is there a command zppy could include to do that automatically for the user? If so, I can make a GitHub issue for that.

See answer above. One command that will extract the month of March, say, from a file containing twelve months (i.e., timesteps) that begins in January, and place it in a separate file is, for example:
`ncks -d time,2 in.nc out.nc
`

**Reply by forsyth2** [reply]
Thanks @czender. I created #313 and #314 to follow up further on these issues.
